File: server/camcops_server/tasks/irac.py
#!/usr/bin/env python

"""
camcops_server/tasks/irac.py

===============================================================================

    Copyright (C) 2012-2019 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.

===============================================================================

"""

import cardinal_pythonlib.rnc_web as ws

from sqlalchemy.sql.schema import Column
from sqlalchemy.sql.sqltypes import Integer, UnicodeText

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_html import tr_qa
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_sqla_coltypes import (
    CamcopsColumn,
    ZERO_TO_TWO_CHECKER,
)
from camcops_server.cc_modules.cc_task import (
    get_from_dict,
    Task,
    TaskHasPatientMixin,
)


# =============================================================================
# IRAC
# =============================================================================

class Irac(TaskHasPatientMixin, Task):
    """
    Server implementation of the IRAC task.
    """
    __tablename__ = "irac"
    shortname = "IRAC"

    aim = Column(
        "aim", UnicodeText,
        comment="Main aim of the contact"
    )
    achieved = CamcopsColumn(
        "achieved", Integer,
        permitted_value_checker=ZERO_TO_TWO_CHECKER,
        comment="Was the aim achieved? (0 not, 1 partially, 2 fully)"
    )

    TASK_FIELDS = ["aim", "achieved"]

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Identify and Rate the Aim of the Contact")

    def is_complete(self) -> bool:
        return (self.all_fields_not_none(self.TASK_FIELDS) and
                self.field_contents_valid())

    def get_achieved_text(self, req: CamcopsRequest) -> str:
        achieveddict = {
            None: None,
            0: self.wxstring(req, "achieved_0"),
            1: self.wxstring(req, "achieved_1"),
            2: self.wxstring(req, "achieved_2"),
        }
        return get_from_dict(achieveddict, self.achieved)

    def get_task_html(self, req: CamcopsRequest) -> str:
        if self.achieved is not None:
            achieved = f"{self.achieved}. {self.get_achieved_text(req)}"
        else:
            achieved = None
        return f"""
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {self.get_is_complete_tr(req)}
                </table>
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="50%">Question</th>
                    <th width="50%">Answer</th>
                </tr>
                {tr_qa(self.wxstring(req, "q_aim"), ws.webify(self.aim))}
                {tr_qa(self.wxstring(req, "q_achieved"), achieved)}
            </table>
        """

File: tablet_qt/tasks/irac.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#include "irac.h"

#include "lib/stringfunc.h"
#include "maths/mathfunc.h"
#include "questionnairelib/namevaluepair.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/qumcq.h"
#include "questionnairelib/qutext.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"
using mathfunc::noneNull;
using stringfunc::standardResult;
using stringfunc::strnum;

const QString Irac::IRAC_TABLENAME("irac");

const QString AIM("aim");
const QString ACHIEVED("achieved");

void initializeIrac(TaskFactory& factory)
{
    static TaskRegistrar<Irac> registered(factory);
}

Irac::Irac(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, IRAC_TABLENAME, false, false, false)  // ... anon, clin, resp
{
    addField(AIM, QMetaType::fromType<QString>());
    addField(ACHIEVED, QMetaType::fromType<int>());

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}

// ============================================================================
// Class info
// ============================================================================

QString Irac::shortname() const
{
    return "IRAC";
}

QString Irac::longname() const
{
    return tr("Identify and Rate the Aim of the Contact");
}

QString Irac::description() const
{
    return tr("Clinician-specified aim of contact, and whether aim achieved.");
}

// ============================================================================
// Instance info
// ============================================================================

bool Irac::isComplete() const
{
    return noneNull(values({AIM, ACHIEVED}));
}

QStringList Irac::summary() const
{
    return QStringList{
        fieldSummary(AIM, xstring("q_aim")) + ".",
        standardResult(xstring("q_achieved"), getAchievedText()),
    };
}

QStringList Irac::detail() const
{
    return completenessInfo() + summary();
}

OpenableWidget* Irac::editor(const bool read_only)
{
    NameValueOptions options_aim;
    for (int i = 1; i <= 10; ++i) {
        const QString s = xstring(strnum("aim_", i));
        options_aim.append(NameValuePair(s, s));
    }
    NameValueOptions options_achieved;
    for (int i = 0; i <= 2; ++i) {
        const QString s = xstring(strnum("achieved_", i));
        options_achieved.append(NameValuePair(s, i));
    }
    QuPagePtr page((new QuPage{
                        (new QuText(xstring("q_aim")))->setBold(),
                        new QuMcq(fieldRef(AIM), options_aim),
                        (new QuText(xstring("q_achieved")))->setBold(),
                        new QuMcq(fieldRef(ACHIEVED), options_achieved),
                    })
                       ->setTitle(longname()));

    auto questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Clinician);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}

// ============================================================================
// Task-specific calculations
// ============================================================================

QString Irac::getAchievedText() const
{
    const QVariant v = value(ACHIEVED);
    const int i = v.toInt();
    if (v.isNull() || i < 0 || i > 2) {
        return "?";
    }
    return xstring(strnum("achieved_", i));
}
